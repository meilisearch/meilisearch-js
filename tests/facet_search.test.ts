import { expect, test, describe, beforeAll, afterAll } from "vitest";
import {
  clearAllIndexes,
  config,
  getClient,
} from "./utils/meilisearch-test-utils.js";

const index = {
  uid: "movies_test",
};

const dataset = [
  {
    id: 123,
    title: "Pride and Prejudice",
    genres: ["romance", "action"],
  },
  {
    id: 456,
    title: "Le Petit Prince",
    genres: ["adventure", "comedy"],
  },
  {
    id: 2,
    title: "Le Rouge et le Noir",
    genres: "romance",
  },
  {
    id: 1,
    title: "Alice In Wonderland",
    genres: ["adventure"],
  },
];

describe.each([
  { permission: "Master" },
  { permission: "Admin" },
  { permission: "Search" },
])("Test on POST search", ({ permission }) => {
  beforeAll(async () => {
    await clearAllIndexes(config);
    const client = await getClient("Master");
    const newFilterableAttributes = ["genres", "title"];
    await client.createIndex({ uid: index.uid });
    await client.index(index.uid).updateSettings({
      filterableAttributes: newFilterableAttributes,
    });
    await client.index(index.uid).addDocuments(dataset).waitTask();
  });

  test(`${permission} key: basic facet value search`, async () => {
    const client = await getClient(permission);

    const params = {
      facetQuery: "a",
      facetName: "genres",
    };
    const response = await client.index(index.uid).searchForFacetValues(params);

    expect(response).toMatchSnapshot();
  });

  test(`${permission} key: facet value search with no facet query`, async () => {
    const client = await getClient(permission);

    const params = {
      facetName: "genres",
    };
    const response = await client.index(index.uid).searchForFacetValues(params);

    expect(response).toMatchSnapshot();
  });

  test(`${permission} key: facet value search with filter`, async () => {
    const client = await getClient(permission);

    const params = {
      facetName: "genres",
      facetQuery: "a",
      filter: ["genres = action"],
    };

    const response = await client.index(index.uid).searchForFacetValues(params);

    expect(response).toMatchSnapshot();
  });

  test(`${permission} key: facet value search with search query`, async () => {
    const client = await getClient(permission);

    const params = {
      facetName: "genres",
      facetQuery: "a",
      q: "Alice",
    };
    const response = await client.index(index.uid).searchForFacetValues(params);

    // @TODO: This is flaky, processingTimeMs is not guaranteed
    expect(response).toMatchSnapshot();
  });

  test(`${permission} key: facet value search with exhaustive facet count`, async () => {
    const client = await getClient(permission);

    const params = {
      facetName: "genres",
      facetQuery: "a",
      q: "Alice",
      exhaustiveFacetCount: true,
    };
    const response = await client.index(index.uid).searchForFacetValues(params);

    // @TODO: This is flaky, processingTimeMs is not guaranteed
    expect(response).toMatchSnapshot();
  });
});

afterAll(() => {
  return clearAllIndexes(config);
});
