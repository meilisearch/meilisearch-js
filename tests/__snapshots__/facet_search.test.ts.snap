// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test on POST search > Admin key: basic facet value search 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "action",
    },
    {
      "count": 2,
      "value": "adventure",
    },
  ],
  "facetQuery": "a",
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Admin key: facet value search with exhaustive facet count 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "adventure",
    },
  ],
  "facetQuery": "a",
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Admin key: facet value search with filter 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "action",
    },
  ],
  "facetQuery": "a",
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Admin key: facet value search with no facet query 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "action",
    },
    {
      "count": 2,
      "value": "adventure",
    },
    {
      "count": 1,
      "value": "comedy",
    },
    {
      "count": 2,
      "value": "romance",
    },
  ],
  "facetQuery": null,
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Admin key: facet value search with search query 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "adventure",
    },
  ],
  "facetQuery": "a",
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Master key: basic facet value search 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "action",
    },
    {
      "count": 2,
      "value": "adventure",
    },
  ],
  "facetQuery": "a",
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Master key: facet value search with exhaustive facet count 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "adventure",
    },
  ],
  "facetQuery": "a",
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Master key: facet value search with filter 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "action",
    },
  ],
  "facetQuery": "a",
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Master key: facet value search with no facet query 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "action",
    },
    {
      "count": 2,
      "value": "adventure",
    },
    {
      "count": 1,
      "value": "comedy",
    },
    {
      "count": 2,
      "value": "romance",
    },
  ],
  "facetQuery": null,
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Master key: facet value search with search query 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "adventure",
    },
  ],
  "facetQuery": "a",
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Search key: basic facet value search 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "action",
    },
    {
      "count": 2,
      "value": "adventure",
    },
  ],
  "facetQuery": "a",
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Search key: facet value search with exhaustive facet count 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "adventure",
    },
  ],
  "facetQuery": "a",
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Search key: facet value search with filter 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "action",
    },
  ],
  "facetQuery": "a",
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Search key: facet value search with no facet query 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "action",
    },
    {
      "count": 2,
      "value": "adventure",
    },
    {
      "count": 1,
      "value": "comedy",
    },
    {
      "count": 2,
      "value": "romance",
    },
  ],
  "facetQuery": null,
  "processingTimeMs": 0,
}
`;

exports[`Test on POST search > Search key: facet value search with search query 1`] = `
{
  "facetHits": [
    {
      "count": 1,
      "value": "adventure",
    },
  ],
  "facetQuery": "a",
  "processingTimeMs": 0,
}
`;
